odoo/tools/date_utils.py

```diff
@@ -178,3 +177,45 @@ def json_default(obj):
             return fields.Datetime.to_string(obj)
         return fields.Date.to_string(obj)
     return ustr(obj)
+
+def date_range(start, end, step=relativedelta(months=1)):
+    """Date range generator with a step interval.
+
+    :param start: A datetime. begining date of the range.
+    :param end: A datetime. ending date of the range.
+    :param step: A relativedelta. interval of the range.
+    :return: A Generator[datetime, None, None].
+    """
+
+    are_naive = start.tzinfo is None and end.tzinfo is None
+    are_utc = start.tzinfo == pytz.utc and end.tzinfo == pytz.utc
+
+    # Cases with miscellenous timezone are more complexe because of DST.
+    are_others = start.tzinfo and end.tzinfo and not are_utc
+
+    if are_others:
+        if start.tzinfo.zone != end.tzinfo.zone:
+            raise ValueError("Timezones of start argument and end argument seem inconsistent")
+
+    if not are_naive and not are_utc and not are_others:
+        raise ValueError("Timezones of start argument and end argument mismatch")
+
+    if start == end:
+        raise ValueError("start and end can't be equal")
```

Comment:
don't block this edge case, if the implementation works, it seems valid + fix corresponding test